### PR TITLE
add `OpInfo` for `torch.diagflat`

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65681
* #68419
* #69212
* #67079
* #67023
* #67022
* #69211
* __->__ #65680
* #65469

cc @mruberry

Differential Revision: [D31730001](https://our.internmc.facebook.com/intern/diff/D31730001)

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -5696,6 +5696,16 @@ def sample_inputs_kl_div(op_info, device, dtype, requires_grad, **kwargs):
         )
     return sample_inputs
 
+def sample_inputs_diagflat(op_info, device, dtype, requires_grad, **kwargs):
+    make_input = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    return [
+        SampleInput(make_input((2,))),
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:5703]: Could we test a scalar tensor? (Surprisingly this does work):
```
x = torch.tensor(2.)
torch.diagflat(x)
```

Comment (zou3519) [reply torch/testing/_internal/common_methods_invocations.py:5703]: Also this:
```
x = torch.tensor([])
torch.diagflat(x)
```

**Review by zou3519 — COMMENTED** [review]
This looks pretty good, I added some suggestions for edge cases

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:5703]: I'm not sure that we should. The documentation [only mentions behavior for n-dimensional tensors with `n >= 1`](https://pytorch.org/docs/master/generated/torch.diagflat.html#torch.diagflat):

> - If `input` is a vector (1-D tensor), then returns a 2-D square tensor with the elements of `input` as the diagonal.
>
> - If `input` is a tensor with more than one dimension, then returns a 2-D tensor with diagonal elements equal to a flattened `input`.

Adding the scalar or empty cases you suggested would mean we test undocumented behavior.

Comment (zou3519) [reply torch/testing/_internal/common_methods_invocations.py:5703]: There's a lot of undocumented behavior that we do test in OpInfo, especially around scalar tensors. For example, torch.sum's `https://pytorch.org/docs/stable/generated/torch.sum.html` documentation doesn't mention what happens with the following:
```
x = torch.tensor(3.)
x.sum(0)
```
but this strangely works. What do you think?

Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:5703]: IMO we should only test intended behavior and I wouldn't take the fact that something works as intended behavior. I would consider the documentation the "ground truth" for the behavior. So either we document the scalar / degenerate behavior or we don't test it.

The fact that scalar inputs work might very well be a side effect of the current implementation. If we now test for it and in the future change the implementation in a way that no longer supports scalars, we may find us in a situation where we are trying to keep BC for behavior that was unintended in the first place.

Not a hill I want to die on. If you want to have scalar / degenerate sample inputs, I'll add them.

Comment (zou3519) [reply torch/testing/_internal/common_methods_invocations.py:5703]: > Not a hill I want to die on. If you want to have scalar / degenerate sample inputs, I'll add them.

Let's keep this PR as-is and we can push the discussion scalar/degenerate sample inputs to later

**Review by zou3519 — APPROVED** [review]

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -14121,6 +14131,15 @@ def ref_pairwise_distance(input1, input2):
             ),
         ),
     ),
+    OpInfo(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:14134]: Ug, this file has gotten too big and it's time for a refactor. But not an issue for this PR, of course

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:7529]: Whether we test something or not doesn't impact our decisions re: BC-breaking changes. If we support something, accidentally or not, it's a feature.

Let's just add one such sample test.

**Review by mruberry — APPROVED** [review]
Let's just add the one additional test case that @zou3519 suggests then ping me to merge this, @pmeier

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:14138]: IIRC there was some effort from @antocuni to do that. Any progress on that?

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:14138]: His work is good it just requires syncing on my end

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31730001).

**Pull request closed without merging**